### DISCUSSION TITLE
brew update shows deleted software

### DISCUSSION THREAD

**Discussion by Kha1r0** [Homebrew/discussions#5699] [Everyday usage] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.4.3
ORIGIN: https://github.com/Homebrew/brew
HEAD: 91f3671b16fe62121364c465c14e0dfd088d7ead
Last commit: 2 days ago
Core tap JSON: 29 Oct 22:40 UTC
Core cask tap JSON: 29 Oct 22:40 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 3.3.5 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.5/bin/ruby
CPU: 10-core 64-bit arm_blizzard_avalanche
Clang: 16.0.0 build 1600
Git: 2.47.0 => /opt/homebrew/bin/git
Curl: 8.7.1 => /usr/bin/curl
macOS: 15.0.1-arm64
CLT: 16.0.0.0.1.1724870825
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

<img width="497" alt="Screenshot 2024-10-29 at 23 37 58" src="https://github.com/user-attachments/assets/76382a75-7635-435a-84b6-dd89f84e2e2b">

After running brew update, i get the message that icu4c is deleted from the repository. Does that mean my device is running unmaintained, possibly insecure software? Do i have reason to be worried?

**Comment by p-linnane** [comment, marked as answer]
No, we just versioned `icu4c`. If you run `brew info icu4c` you will see it is now an alias to `icu4c@75`.

**Reply by Kha1r0** [reply]
Thanks! How do i know I‘m safe if i see a message like this in the future? Sorry if this is a stupid question.

**Reply by p-linnane** [reply]
If you run `brew doctor` it will let you know if you are using discontinued software:

```
Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  cowsay
```

**Reply by Kha1r0** [reply]
Thank you! ❤️

**Comment by tobbi** [comment]
@p-linnane For me it also automatically uninstalled icu4c with that message. However, there were formulas that depended on icu4c that now failed to build. 

In my case, libxml2:

```
==> Upgrading libxml2
  2.13.4_2 -> 2.13.4_3
==> ./configure --sysconfdir=/usr/local/etc --disable-silent-rules --with-history --with-http --with-icu --without-python --without-lzma
==> make install
Last 15 lines from /Users/tobiasmarkus/Library/Logs/Homebrew/libxml2/02.make:
      _xmlFindExtraHandler in libxml2_la-encoding.o
      _xmlCharEncCloseFunc in libxml2_la-encoding.o
      _openIcuConverter in libxml2_la-encoding.o
  "_ucnv_convertEx_76", referenced from:
      _xmlUconvWrapper in libxml2_la-encoding.o
  "_ucnv_open_76", referenced from:
      _openIcuConverter in libxml2_la-encoding.o
  "_ucnv_setFromUCallBack_76", referenced from:
      _openIcuConverter in libxml2_la-encoding.o
  "_ucnv_setToUCallBack_76", referenced from:
      _openIcuConverter in libxml2_la-encoding.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [libxml2.la] Error 1
```

**Reply by tobbi** [reply]
I understand that obviously installing icu4c 0.76 will probably fix it, but: Could this not have been done without breaking installed formulas?

**Reply by p-linnane** [reply]
Given that you are building from source to upgrade, you are most likely running an unsupported configuration. Building from source is not officially supported.

**Reply by tobbi** [reply]
This probably has something to do with me running Mac OS Mojave on an unsupported MacBook Pro using Open Core Patcher.

**Reply by p-linnane** [reply]
Yeah, that is the cause of your issue unfortunately. `icu4c` was upgraded, and we the oldest bottles we build are for macOS Ventura. You will unfortunately need to debug it.

**Reply by meisterfry** [reply]
@tobbi Did you manage to debug it? I currently get the exact same Error when i try to build libxml2 from source on macos Monterey. Already tried installing icu4c@76, completely uninstalling libxml2 und reinstalling all it's dependencys, but i couldn't get it to build successfully.

**Reply by tobbi** [reply]
Unfortunately not. I don't know whether homebrew is not linking in part of the ICU dependencies

**Reply by meisterfry** [reply]
That's a real shame. The only thing i can rule out is the possibility that it has something to do with your OCP configuration, since it also happens on my Macbook Air with native Monterey. Unfortunately my technical skills are very limited, i tried some very simple debugging steps and asked different LLMs, but the only answers provided are things like setting the LD/CPPFLAGS for icu4c@76 which i had obviously already done. Have you tried uninstalling homebrew completely and reinstall everything from scratch?

**Reply by EricFromCanada** [reply]
The missing symbols are in `libicuuc.76.dylib`:
```
$ nm -gU /opt/homebrew/opt/icu4c@76/lib/libicuuc.76.dylib | grep _ucnv_setToUCallBack_76
0000000000065868 T _ucnv_setToUCallBack_76
```
On ARM w/ Sonoma this library is linked to:
```
$ brew linkage libxml2
System libraries:
  /usr/lib/libSystem.B.dylib
  /usr/lib/libiconv.2.dylib
  /usr/lib/libncurses.5.4.dylib
  /usr/lib/libz.1.dylib
Homebrew libraries:
  /opt/homebrew/opt/icu4c@76/lib/libicudata.76.dylib (icu4c@76)
  /opt/homebrew/opt/icu4c@76/lib/libicui18n.76.dylib (icu4c@76)
  /opt/homebrew/opt/icu4c@76/lib/libicuuc.76.dylib (icu4c@76)
  /opt/homebrew/Cellar/libxml2/2.13.4_4/lib/libxml2.2.dylib (libxml2)
  /opt/homebrew/opt/readline/lib/libhistory.8.dylib (readline)
  /opt/homebrew/opt/readline/lib/libreadline.8.dylib (readline)
```
which we can see when building from source:
```
/bin/sh ./libtool  --tag=CC   --mode=link clang -pedantic -Wall -Wextra -Wshadow -Wpointer-arith 
…
libxml2_la-xpointer.lo -liconv -L/opt/homebrew/Cellar/icu4c/74.2/lib -licui18n -licuuc -licudata
```
while on Intel w/ Monterey, `-licuuc -licudata` are missing:
```
/bin/sh ./libtool  --tag=CC   --mode=link clang -pedantic -Wall -Wextra -Wshadow -Wpointer-arith
…
libxml2_la-xpointer.lo -L/usr/lib -lz -L/usr/local/Cellar/icu4c@76/76.1_1/lib -licui18n  -liconv
```

**Reply by ZhongRuoyu** [reply]
I encountered this a while back too. I haven't had the chance to submit the fix upstream, but I believe this is because [libxml2's pkg-config check for icu](https://github.com/GNOME/libxml2/blob/aaecdc92e237d0873578f26773085108cfb972e6/configure.ac#L985-L993) is missing `icu-uc`:

```autoconf
    # Try pkg-config first so that static linking works.
    # If this succeeeds, we ignore the WITH_ICU directory.
    PKG_CHECK_MODULES([ICU], [icu-i18n], [
        WITH_ICU=1; XML_PC_REQUIRES="${XML_PC_REQUIRES} icu-i18n"
        m4_ifdef([PKG_CHECK_VAR],
            [PKG_CHECK_VAR([ICU_DEFS], [icu-i18n], [DEFS])])
        if test "x$ICU_DEFS" != "x"; then
            ICU_CFLAGS="$ICU_CFLAGS $ICU_DEFS"
        fi],[:])
```

 We were able to build this in Homebrew CI because `pkg-config` happened to be absent in the dep tree.

Their [CMake counterpart](https://github.com/GNOME/libxml2/blob/aaecdc92e237d0873578f26773085108cfb972e6/CMakeLists.txt#L120-L122) is good:

```cmake
if(LIBXML2_WITH_ICU)
    find_package(ICU REQUIRED COMPONENTS data i18n uc)
endif()
```

**Reply by EricFromCanada** [reply]
Just confirmed that first running `brew uninstall pkg-config` allows `brew install libxml2` to succeed on platforms without bottles. Thanks!

**Reply by ZhongRuoyu** [reply]
Upstreamed a patch: https://gitlab.gnome.org/GNOME/libxml2/-/merge_requests/289. Also fixing this at https://github.com/Homebrew/homebrew-core/pull/197591.

**Reply by ZhongRuoyu** [reply]
Update: Homebrew's `libxml2` has been fixed. It's now possible to build `libxml2` without the `pkg-config` workaround.
